File: app/src/main/java/com/brizbee/Brizbee/Mobile/PunchInTaskIdActivity.java
package com.brizbee.Brizbee.Mobile;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import com.android.volley.AuthFailureError;
import com.android.volley.DefaultRetryPolicy;
import com.android.volley.NetworkResponse;
import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.RetryPolicy;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.google.zxing.integration.android.IntentIntegrator;
import com.google.zxing.integration.android.IntentResult;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

public class PunchInTaskIdActivity extends AppCompatActivity implements View.OnClickListener
{
    private Button buttonScan;
    private TextView editTaskNumber;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_punch_in_task_id);

        // Get references from layouts
        buttonScan = findViewById(R.id.buttonScan);
        editTaskNumber = findViewById(R.id.editTaskNumber);

        // Set the click listener
        buttonScan.setOnClickListener(this);

        // Focus on the task id
        if(editTaskNumber.requestFocus())
        {
            getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
        }
    }

    public void onClick(View v)
    {
        // Respond to clicks
        if(v.getId() == R.id.buttonScan)
        {
            //scan
            IntentIntegrator scanIntegrator = new IntentIntegrator(this);
            scanIntegrator.initiateScan();
        }
    }

    public void onActivityResult(int requestCode, int resultCode, Intent intent)
    {
        // Retrieve the scan result
        IntentResult scanningResult = IntentIntegrator.parseActivityResult(requestCode, resultCode, intent);
        if (scanningResult != null)
        {
            String scanContent = scanningResult.getContents();
            String scanFormat = scanningResult.getFormatName(); // barcode type
            editTaskNumber.setText(""); // clear the text
            editTaskNumber.setText(scanContent); // set the scanned value
            confirm(); // Verify that the task number is valid
        }
        else
        {
            showDialog("No task number could be scanned, please try again.");
        }
    }

    public void onCancelClick(View view)
    {
        final Intent intent = new Intent(this, StatusActivity.class);
        startActivity(intent);
        finish(); // prevents going back
    }

    public void onContinueClick(View view)
    {
        confirm();
    }

    public void confirm()
    {
        // Lookup the task number
        final Intent intent = new Intent(this, PunchInConfirmActivity.class);
        final AlertDialog.Builder builder = new AlertDialog.Builder(this);
        EditText editTaskNumber = findViewById(R.id.editTaskNumber);

        // Instantiate the RequestQueue
        String url = String.format("https://brizbee.gowitheast.com/odata/Tasks?$expand=Job($expand=Customer)&$filter=Number eq '%s'", editTaskNumber.getText());

        JsonObjectRequest jsonRequest = new JsonObjectRequest
                (Request.Method.GET, url, null, new Response.Listener<JSONObject>()
                {
                    @Override
                    public void onResponse(JSONObject response)
                    {
                        try {
                            JSONArray value = response.getJSONArray("value");

                            if (value.length() > 0) {
                                JSONObject first = value.getJSONObject(0);

                                // Pass the task as a string
                                intent.putExtra("task", first.toString());
                                startActivity(intent);
                            } else {
                                // Notify the user that the task number does not exist
                                showDialog("That's not a valid task number, please try again.");
                            }
                        } catch (JSONException e) {
                            showDialog("Could not understand the response from the server, please try again.");
                        }
                    }
                },
                        new Response.ErrorListener()
                {
                    @Override
                    public void onErrorResponse(VolleyError error)
                    {
                        NetworkResponse response = error.networkResponse;
                        switch (response.statusCode)
                        {
                            default:
                                showDialog("Could not reach the server, please try again.");
                                break;
                        }
                    }
                }) {
            @Override
            public Map<String, String> getHeaders() throws AuthFailureError
            {
                HashMap<String, String> headers = new HashMap<String, String>();
                headers.put("Content-Type", "application/json");

                String authExpiration = ((MyApplication) getApplication()).getAuthExpiration();
                String authToken = ((MyApplication) getApplication()).getAuthToken();
                String authUserId = ((MyApplication) getApplication()).getAuthUserId();

                if (authExpiration != null && !authExpiration.isEmpty() &&
                        authToken != null && !authToken.isEmpty() &&
                        authUserId != null && !authUserId.isEmpty()) {
                    headers.put("AUTH_EXPIRATION", authExpiration);
                    headers.put("AUTH_TOKEN", authToken);
                    headers.put("AUTH_USER_ID", authUserId);
                }

                return headers;
            }
        };

        int socketTimeout = 10000;
        RetryPolicy policy = new DefaultRetryPolicy(socketTimeout, 3, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT);
        jsonRequest.setRetryPolicy(policy);

        // Add the request to the RequestQueue
        MySingleton.getInstance(this).addToRequestQueue(jsonRequest);
    }

    private void showDialog(String message)
    {
        // Build a dialog with the given message to show the user
        androidx.appcompat.app.AlertDialog.Builder builder = new androidx.appcompat.app.AlertDialog.Builder(this);
        builder.setMessage(message)
                .setPositiveButton("OK", new DialogInterface.OnClickListener()
                {
                    public void onClick(DialogInterface dialog, int id)
                    {
                        dialog.dismiss();
                    }
                });
        androidx.appcompat.app.AlertDialog dialog = builder.create();
        dialog.show();
    }
}

File: app/src/main/java/com/brizbee/Brizbee/Mobile/PunchInTaskIdActivity.kt
//
//  PunchInTaskIdActivity.kt
//  BRIZBEE Mobile for Android
//
//  Copyright © 2021 East Coast Technology Services, LLC
//
//  This file is part of BRIZBEE Mobile for Android.
//
//  BRIZBEE Mobile for Android is free software: you can redistribute
//  it and/or modify it under the terms of the GNU General Public
//  License as published by the Free Software Foundation, either
//  version 3 of the License, or (at your option) any later version.
//
//  BRIZBEE Mobile for Android is distributed in the hope that it will
//  be useful, but WITHOUT ANY WARRANTY; without even the implied
//  warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//  See the GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with BRIZBEE Mobile for Android.
//  If not, see <https://www.gnu.org/licenses/>.
//

package com.brizbee.Brizbee.Mobile

import android.app.Activity
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.content.Intent
import android.widget.EditText
import com.android.volley.toolbox.JsonObjectRequest
import org.json.JSONException
import android.view.View
import android.view.WindowManager
import android.widget.Button
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.camera.core.ExperimentalGetImage
import com.android.volley.*
import java.util.HashMap
import kotlin.concurrent.thread

@ExperimentalGetImage
class PunchInTaskIdActivity : AppCompatActivity() {
    private var buttonScan: Button? = null
    private var editTaskNumber: EditText? = null
    var progressDialog: AlertDialog? = null

    companion object {
        val TAG = PunchInTaskIdActivity::class.qualifiedName
    }

    private val startForResult = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult ->
        if (result.resultCode == Activity.RESULT_OK) {
            val scanContent = result.data?.extras?.get("BarCodeValue").toString()
            editTaskNumber?.setText(scanContent)

            // Verify that the barcode is valid.
            thread(start = true) {
                confirm()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_punch_in_task_id)

        // Get references from layouts.
        buttonScan = findViewById(R.id.buttonScan)
        editTaskNumber = findViewById(R.id.editTaskNumber)

        // Respond to button click.
        buttonScan?.setOnClickListener {
            // Start the barcode scanner.
            startForResult.launch(Intent(this, BarcodeScanActivity::class.java))
        }

        // Focus on the task number.
        if (editTaskNumber?.requestFocus() == true) {
            window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE)
        }
    }

    @Suppress("UNUSED_PARAMETER")
    fun onCancelClick(view: View?) {
        val intent = Intent(this, StatusActivity::class.java)
        startActivity(intent)
        finish()
    }

    @Suppress("UNUSED_PARAMETER")
    fun onContinueClick(view: View?) {
        thread(start = true) {
            confirm()
        }
    }

    private fun confirm() {
        runOnUiThread {
            // Prepare a progress dialog.
            val builder = AlertDialog.Builder(this)
            builder.setMessage("Working")
            builder.setCancelable(false)
            progressDialog = builder.create()
            progressDialog?.setCancelable(false)
            progressDialog?.setCanceledOnTouchOutside(false)
            progressDialog?.show()
        }

        // Lookup the task number.
        val editTaskNumber = findViewById<EditText>(R.id.editTaskNumber)

        // Instantiate the RequestQueue.
        val url = "${(application as MyApplication).baseUrl}/api/Kiosk/SearchTasks?taskNumber=${editTaskNumber.text}"

        val request: JsonObjectRequest = object : JsonObjectRequest(Method.GET, url, null,
            { response ->
                runOnUiThread {
                    try {
                        if (response.getJSONObject("job").getString("status").uppercase() != "OPEN") {
                            // Notify the user that the project is not open
                            showDialog("The project for that task number is not open, please try again.")
                            progressDialog?.dismiss()
                        } else {
                            val intent = Intent(this, PunchInConfirmActivity::class.java)

                            // Pass the task as a string
                            intent.putExtra("task", response.toString())
                            startActivity(intent)
                        }
                    } catch (e: JSONException) {
                        runOnUiThread {
                            progressDialog?.dismiss()
                            showDialog("Could not understand the response from the server, please try again.")
                        }
                    }
                }
            }, { error ->
                runOnUiThread {
                    val response = error.networkResponse
                    when (response.statusCode) {
                        404 -> {
                            progressDialog?.dismiss()
                            showDialog("That's not a valid task number, please try again.")
                        }
                        else -> {
                            progressDialog?.dismiss()
                            showDialog("Could not reach the server, please try again.")
                        }
                    }
                }
            }) {
                override fun getHeaders(): Map<String, String> {
                    return (application as MyApplication).authHeaders
                }
            }

        // Increase number of retries because we may be on a poor connection.
        request.retryPolicy = DefaultRetryPolicy(10000, 3, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)

        // Add the request to the RequestQueue.
        request.tag = TAG
        MySingleton.getInstance(this).addToRequestQueue(request)
    }

    private fun showDialog(message: String) {
        runOnUiThread {
            // Build a dialog with the given message to show the user
            val builder = AlertDialog.Builder(this)
            builder.setMessage(message)
                .setPositiveButton(
                    "OK"
                ) { dialog, _ -> dialog.dismiss() }
            val dialog = builder.create()
            dialog.show()
        }
    }
}
